sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -87,20 +87,41 @@ RT::PiProgram ProgramManager::getBuiltOpenCLProgram(OSModuleHandle M,
   std::shared_ptr<context_impl> Ctx = getSyclObjImpl(Context);
   std::map<OSModuleHandle, RT::PiProgram> &CachedPrograms =
       Ctx->getCachedPrograms();
-  auto It = CachedPrograms.find(M);
-  if (It != CachedPrograms.end())
-    return It->second;
+
+  {
+    std::lock_guard<std::mutex> CacheLock(Ctx->getProgramKernelCacheMutex());
+
+    auto It = CachedPrograms.find(M);
+    if (It != CachedPrograms.end())
+      return It->second;
+  }
+
+  // TODO improvement to employ after code-split feature is done:
+  // - put atomic pointer in program cache as data with nullptr before
+  //   the program is actually built
+  // - any thread which will find nullptr in cache will wait until the pointer
+  //   is no null anymore
 
   DeviceImage *Img = nullptr;
   using PiProgramT = remove_pointer_t<RT::PiProgram>;
   unique_ptr_class<PiProgramT, decltype(RT::piProgramRelease)> ProgramManaged(
       loadProgram(M, Context, &Img), RT::piProgramRelease);
 
   build(ProgramManaged.get(), Img->BuildOptions);
-  RT::PiProgram Program = ProgramManaged.release();
-  CachedPrograms[M] = Program;
 
-  return Program;
+  {
+    std::lock_guard<std::mutex> CacheLock(Ctx->getProgramKernelCacheMutex());
+
+    auto It = CachedPrograms.find(M);
```

Comment:
You don't have to lookup twice, the iterator stays valid (for std::map), although you should use insert or operator[] to make sure that an empty element is created.